sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/MethodSignatureHelper.cs

```diff
@@ -19,25 +19,28 @@
  */
 
 using System;
+using System.Linq;
 using Microsoft.CodeAnalysis;
-using Microsoft.CodeAnalysis.CSharp;
 using Microsoft.CodeAnalysis.CSharp.Syntax;
 
 namespace SonarAnalyzer.Helpers
 {
     public static class MethodSignatureHelper
     {
-        public static Func<MethodSignature, bool> IsSame<TSymbolType>(SimpleNameSyntax identifierName, SemanticModel semanticModel)
+        public static bool IsMatch<TSymbolType>(SimpleNameSyntax identifierName, SemanticModel semanticModel,
+            Lazy<TSymbolType> symbolFetcher, MethodSignature[] methods)
             where TSymbolType : class, ISymbol
         {
+            if (identifierName == null)
+            {
+                return false;
+            }
+
             var identifierText = identifierName.Identifier.ValueText;
-            var identifierSymbol = new Lazy<TSymbolType>(() => semanticModel.GetSymbolInfo(identifierName).Symbol as TSymbolType);
-            // This function will be called multiple times for each tracked MethodSignature,
-            // hence we cache as much as possible before.
-            return (methodSignature) =>
-                identifierText == methodSignature.Name &&
-                identifierSymbol.Value != null &&
-                identifierSymbol.Value.ContainingType.Is(methodSignature.ContainingType);
+            return methods.Any(m =>
+                identifierText == m.Name &&
+                symbolFetcher.Value != null &&
+                symbolFetcher.Value.ContainingType.Is(m.ContainingType));
```

Comment:
Other classes that match MethodSignature usually use `symbol.ContainingType.ConstructedFrom.Is(knownType)`... This could be fixed later when we find it, but it is definitely something we should have in mind with generic types.

Reply:
I'll change it now so we don't forget.